### PR TITLE
Simplified signal handling

### LINKED ISSUE #34: on panic, raw terminal mode is not left
## Summary

running into `unimplemented!` or being terminated via `kill -9`, the terminal is stuck in raw mode

## Steps to reproduce

`cargo spellcheck fix -i &; kill -9 $!`

## Expected results

terminal should return to it's previous state

## Actual results

terminal is stuck in raw mode

## Triaging info

* Version: 0.2.1
* Did this used to work? Nope

### LINKED ISSUE #59: Signal thread panic if the `stdout` is redirected
**Overview**

I was willing to use `grep` on the `spellcheck` output and noticed that redirecting output by `>` or `|` will cause panic.
I see that the real output goes to `stderr` so it's not crucial.
I decided just to publish that but it's certainly not critical, even though hard to say why that's raised.

**Reproduction**

```bash
[zhiburt@zhiburt cargo-spellcheck]$ cargo spellcheck | grep ".md"
thread '<unnamed>' panicked at 'Failed to get stdout: Os { code: 25, kind: Other, message: "Inappropriate ioctl for device" }', src/main.rs:87:30
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Expected behavior**

Probably don't having panic is a good one.

**Pplease complete the following information:**
 - System: Arch
 - Obtained: git `4dd8e0666b6c373d04f09fcd7b42a270e7c4fad9`
 - Version: `cargo-spellcheck 0.3.0-alpha.1`

### PR DESCRIPTION
**Pull request by KuabeM**
## What does this PR accomplish?

closes #34 eventually and fixes #59 which was introduced in #58 .

Added the fix for #60 since its just one character..

## Changes proposed by this PR:

The current implementation does not restore the cursor on `SIGINT`, `SIGTERM` and `SIGQUIT`. It also does not work when in interactive mode. To simplify things, I removed the `termios` dep and solely restore the `crossterm` and `std::io::stdout` internal changed in the first place when entered interactive mode.

One drawback: hitting `ctrl+c` in interactive mode leads to exit code `1` instead of `130`. Should we fix this?

### CODE REVIEW

File: src/action/interactive.rs [hunk 1]
```diff
@@ -196,7 +196,9 @@ impl UserPicked {
                 return Ok(Pick::Replacement(bandaid));
             }
             KeyCode::Esc => return Ok(Pick::Quit),
-            KeyCode::Char('c') if modifiers == KeyModifiers::CONTROL => return Ok(Pick::Quit),
+            KeyCode::Char('c') if modifiers == KeyModifiers::CONTROL => {
+                return Err(anyhow!("Received SIGINT"))
```
Comment (drahnr) [inline src/action/interactive.rs:200]: will cause an error code `1`, and not `130` - I think we should exit the process right there

File: src/action/interactive.rs [hunk 2]
```diff
@@ -439,7 +441,9 @@ impl UserPicked {
                 }
                 KeyCode::Char('n') => return Ok(Pick::Skip),
                 KeyCode::Char('j') => return Ok(Pick::Previous),
-                KeyCode::Char('c') if modifiers == KeyModifiers::CONTROL => return Ok(Pick::Quit),
+                KeyCode::Char('c') if modifiers == KeyModifiers::CONTROL => {
+                    return Err(anyhow!("Received SIGINT"))
```
Comment (drahnr) [inline src/action/interactive.rs:445]: CC

File: src/action/mod.rs [hunk 3]
```diff
@@ -197,7 +197,7 @@ impl Action {
         for (_path, suggestions) in suggestions_per_path {
             count += suggestions.len();
             for suggestion in suggestions {
-                eprintln!("{}", suggestion);
```
Comment (drahnr) [inline src/action/mod.rs:200]: :+1:

File: src/main.rs [hunk 4]
```diff
@@ -72,26 +71,15 @@ struct Args {
...
-    let mut termios_stdout = termios::Termios::from_fd(stdout).expect("Failed to get stdout");
-    termios::tcgetattr(stdin, &mut termios_stdin).expect("Failed to get stdin mode");
-    termios::tcgetattr(stdout, &mut termios_stdout).expect("Failed to get stdin mode");
     for s in signals.forever() {
         match s {
             SIGTERM | SIGINT | SIGQUIT => {
-                if on_exit(termios_stdin, stdin).is_err()
-                    || on_exit(termios_stdout, stdout).is_err()
+                if std::io::stdout().queue(crossterm::cursor::Show).is_err()
+                    || crossterm::terminal::disable_raw_mode().is_err()
```
Comment (drahnr) [inline src/main.rs:82]: missing `flush`?

File: src/main.rs [hunk 5]
```diff
@@ -261,7 +249,14 @@ fn main() -> anyhow::Result<()> {
 
     let suggestion_set = checker::check(&combined, &config)?;
 
-    action.run(suggestion_set, &config)
+    match action.run(suggestion_set, &config) {
+        Ok(_) => Ok(()),
+        Err(e) => {
+            crossterm::terminal::disable_raw_mode()?;
+            std::io::stdout().queue(crossterm::cursor::Show)?;
```
Comment (drahnr) [inline src/main.rs:256]: missing `flush`?

File: src/action/interactive.rs (hunk 1 above)
Comment (KuabeM) [reply src/action/interactive.rs:200]: I wasn't sure if it's a good idea to exit right there.

File: src/main.rs [hunk 6]
```diff
@@ -72,19 +70,14 @@ struct Args {
...
-}
 
 fn signal_handler() {
     let signals =
         iterator::Signals::new(vec![SIGTERM, SIGINT, SIGQUIT]).expect("Failed to create Signals");
     for s in signals.forever() {
         match s {
             SIGTERM | SIGINT | SIGQUIT => {
-                if exit().is_err() {
+                if action::interactive::ScopedRaw::restore_terminal().is_err() {
```
Comment (drahnr) [inline src/main.rs:80]: Should this not always cause an exit code of `130`?

File: src/main.rs [hunk 7]
```diff
@@ -72,27 +70,14 @@ struct Args {
...
-    let mut termios_stdout = termios::Termios::from_fd(stdout).expect("Failed to get stdout");
-    termios::tcgetattr(stdin, &mut termios_stdin).expect("Failed to get stdin mode");
-    termios::tcgetattr(stdout, &mut termios_stdout).expect("Failed to get stdin mode");
     for s in signals.forever() {
         match s {
             SIGTERM | SIGINT | SIGQUIT => {
-                if on_exit(termios_stdin, stdin).is_err()
-                    || on_exit(termios_stdout, stdout).is_err()
-                {
+                if action::interactive::ScopedRaw::restore_terminal().is_err() {
```
Comment (drahnr) [inline src/main.rs:80]:
```suggestion
                let _ = action::interactive::ScopedRaw::restore_terminal();
                std::process::exit(130);
```

Comment (drahnr) [reply src/main.rs:80]: Suggestion is incomplete and won't compile.

File: src/main.rs (hunk 6 above)
Comment (KuabeM) [reply src/main.rs:80]: My reasoning was to let the user know that something went wrong and `ctrl+c` was not 'successful'. Thinking about it, I doubt that anyone will check the exit code if it's actually `130` or anything else.. So maybe print a warning on error and exit with 130 anyway?

**Comment by drahnr** [comment]
:boat:

**Pull request merged**